Honour regex in dependsOnMethods

krmahadevan opened the pull request:
Closes: #141

Fixes #141  .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`
- [X] Auto applied styling via `./gradlew autostyleApply`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

**Note:** For more information on contribution guidelines  please make sure you refer our [Contributing](.github/CONTRIBUTING.md) section for detailed set of steps.

juherr commented on testng-core/src/main/java/org/testng/DependencyMap.java:
```diff
@@ -64,26 +65,32 @@ public ITestNGMethod getMethodDependingOn(String methodName, ITestNGMethod fromM
-    throw new TestNGException(
-        "Method \"" + fromMethod + "\" depends on nonexistent method \"" + methodName + "\"");
+  private static boolean methodNameMatch(Pattern pattern, String method) {
+    String justMethodName = method.substring(method.lastIndexOf(".") + 1);
```
Instead of being less precise, why not build the pattern with the class name?

Could you try some samples where there is inner classes and methods with identical names?
I feel unexpected behavior in some cases.

krmahadevan replied:
@juherr 

> Could you try some samples where there is inner classes

I didnt quite understand this part. Can you share a sample that shows what you are talking about ?

krmahadevan replied:
@juherr ping! Can U pls let me know?

juherr replied:
Are we sure that a method with dependency will not found its regex dependency from an inner class?

Class with a test method that dependsOn x*
Inner class with xx test method.
The test from parent should fail because it don't find its dependency.

juherr reviewed (commented):
It need more tests to cover the feature.

krmahadevan commented:
@juherr - I have added the tests. Please take a look and let me know if there's anything additional that you are looking for.

krmahadevan commented:
@juherr - I have addressed all the review comments. There's one bug that I found which seems to exist when you use dependsOnMethods and use a fully qualified method name. Since that defect exists in regular dependsOnMethods usage (without regex), when I fix that defect I will do it for regex as well.

Here's the issue link https://github.com/cbeust/testng/issues/2840

juherr commented:
What do you thing to add the following case:

```java
public class TestClassSample {

  @Test(dependsOnMethods = "test_C[0-9]{7}")
  public void randomTest() {}

  public static class Inner {
    @Test
    public void test_C6390323() {}
  }
}
```

`randomTest` is supposed to fail here because `test_C6390323` is out of scope.

krmahadevan commented:
@juherr - I have added that test as well. Please check.
